### PR TITLE
Add Partner Schools and Partner Providers

### PR DESCRIPTION
**Pull request by JamieCleare2525**
## Context

- Providers require the ability to assign and remove a network of partner schools.
- Schools require the ability to assign and remove a network of partner providers.

## Changes proposed in this pull request

(Following the same process as Support Users onboarding Schools and Providers)
- Controllers/Routes/Models/etc. to allow Provider users the ability to add/remove partner schools.
- Controllers/Routes/Models/etc. to allow School users the ability to add/remove partner providers.

## Guidance to review

- Follow the examples shown in the screenshots.

## Link to Trello card

https://trello.com/c/KQQ6Kd5l/205-add-remove-school-to-an-itt-provider-add-remove-itt-provider-to-a-school

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated or [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

## Screenshots

(As Patricia (Provider User)) - Assigning/Removing Partner School

https://github.com/DFE-Digital/itt-mentor-services/assets/17162488/73b263cd-3b5e-4165-be57-88382e75f3c8

https://github.com/DFE-Digital/itt-mentor-services/assets/17162488/7f08993e-4c31-4535-94c7-59b5d832df82

(As Anne (School User)) - Assigning/Removing Partner Provider

https://github.com/DFE-Digital/itt-mentor-services/assets/17162488/4ba81ee3-34db-4746-9c28-40f4022a0148

https://github.com/DFE-Digital/itt-mentor-services/assets/17162488/63b6e221-d5a4-40b9-87c5-f42b5d6983e9

### CODE REVIEW

File: app/forms/placements/partnership_form.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    @school ||= Placements::School.find_by(id: school_id)
+  end
+
+  def as_form_params
+    { "partnership" => { school_id: school.id, provider_id: provider.id } }
+  end
+
+  def persist
+    ActiveRecord::Base.transaction do
+      Placements::Partnership.create!(school:, provider:)
```
Comment (CatalinVoineag) [inline app/forms/placements/partnership_form.rb:23]: Not sure why this needs to be in a base transaction. Usually, I use this when I want to revert everything if multiple saves need to happen.

File: app/views/placements/providers/partner_schools/remove.html.erb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+<% end %>
+
+<div class="govuk-width-container">
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <label class="govuk-label govuk-label--l">
+        <span class="govuk-caption-l"><%= @partner_school.name %></span>
+        <%= t(".are_you_sure") %>
+      </label>
+      <%#= render GovukComponent::WarningTextComponent.new(
```
Comment (CatalinVoineag) [inline app/views/placements/providers/partner_schools/remove.html.erb:14]: Left this here

File: app/views/placements/schools/partner_providers/remove.html.erb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+<% end %>
+
+<div class="govuk-width-container">
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <label class="govuk-label govuk-label--l">
+        <span class="govuk-caption-l"><%= @partner_provider.name %></span>
+        <%= t(".are_you_sure") %>
+      </label>
+      <%#= render GovukComponent::WarningTextComponent.new(
```
Comment (CatalinVoineag) [inline app/views/placements/schools/partner_providers/remove.html.erb:14]: This one as well

**Review by CatalinVoineag — COMMENTED** [review]
There is quite a lot here, overall I think it looks quite good, well done. Just minor things

File: app/views/placements/providers/partner_schools/remove.html.erb (hunk 2 above)
Comment (JamieCleare2525) [reply app/views/placements/providers/partner_schools/remove.html.erb:14]: Those are left there on purpose, as a remind to add the warnings when doing the second part
https://github.com/DFE-Digital/itt-mentor-services/pull/452

File: app/views/placements/schools/partner_providers/remove.html.erb (hunk 3 above)
Comment (JamieCleare2525) [reply app/views/placements/schools/partner_providers/remove.html.erb:14]: Those are left there on purpose, as a remind to add the warnings when doing the second part
https://github.com/DFE-Digital/itt-mentor-services/pull/452

File: app/views/placements/providers/partner_schools/remove.html.erb (hunk 2 above)
Comment (JamieCleare2525) [reply app/views/placements/providers/partner_schools/remove.html.erb:14]: Actually given I've now done the second part, I'll remove the comments

Comment (JamieCleare2525) [reply app/views/placements/providers/partner_schools/remove.html.erb]: resolved in [9c174f8](https://github.com/DFE-Digital/itt-mentor-services/pull/449/commits/9c174f8926178e92f3eab11b1000174939c20150)

File: app/forms/placements/partnership_form.rb (hunk 1 above)
Comment (JamieCleare2525) [reply app/forms/placements/partnership_form.rb]: Resolved in [de46003](https://github.com/DFE-Digital/itt-mentor-services/pull/449/commits/de4600343e40b57cd20ba586b6321aec3fa9a686)

**Review by CatalinVoineag — APPROVED** [review]

File: spec/models/placements/provider_spec.rb [hunk 4]
```diff
@@ -46,6 +46,25 @@
         expect(placements_provider.users).to all(be_a(Placements::User))
       end
     end
+
+    describe "#partnerships" do
+      it { is_expected.to have_many(:partnerships) }
+    end
+
+    describe "#parnter_schools" do
```
Comment (ollietreend) [inline spec/models/placements/provider_spec.rb:54]:
```suggestion
    describe "#partner_schools" do
```

File: spec/models/placements/school_spec.rb [hunk 5]
```diff
@@ -83,6 +83,25 @@
         expect(placements_school.users).to all(be_a(Placements::User))
       end
     end
+
+    describe "#partnerships" do
+      it { is_expected.to have_many(:partnerships) }
+    end
+
+    describe "#parnter_providers" do
```
Comment (ollietreend) [inline spec/models/placements/school_spec.rb:91]:
```suggestion
    describe "#partner_providers" do
```

File: spec/policies/placements/partnership_policy_spec.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  let(:placements_provider) { create(:placements_provider) }
+  let!(:placements_partnership) do
+    create(
+      :placements_partnership,
+      school: placements_school,
+      provider: placements_provider,
+    )
+  end
+
+  permissions :destroy? do
```
Comment (ollietreend) [inline spec/policies/placements/partnership_policy_spec.rb:16]: Do you also need to cover the `:remove?` permission? Or do you think `:destroy?` is sufficient here?

**Review by ollietreend — COMMENTED** [review]
Nice work @JamieCleare2525 😍 this is looking sharp!

Really nice attention to detail. And I also appreciate the odd little fix here & there which you slipped in too.

File: spec/policies/placements/partnership_policy_spec.rb (hunk 6 above)
Comment (JamieCleare2525) [reply spec/policies/placements/partnership_policy_spec.rb]: resolved in [3602283](https://github.com/DFE-Digital/itt-mentor-services/pull/449/commits/3602283996a7608085d29cdac991157fd989e79d)

File: spec/models/placements/school_spec.rb (hunk 5 above)
Comment (JamieCleare2525) [reply spec/models/placements/school_spec.rb]: resolved in [3602283](https://github.com/DFE-Digital/itt-mentor-services/pull/449/commits/3602283996a7608085d29cdac991157fd989e79d)

File: spec/models/placements/provider_spec.rb (hunk 4 above)
Comment (JamieCleare2525) [reply spec/models/placements/provider_spec.rb]: resolved in [3602283](https://github.com/DFE-Digital/itt-mentor-services/pull/449/commits/3602283996a7608085d29cdac991157fd989e79d)

**Review by ollietreend — APPROVED** [review]

**Pull request merged**